sherpa-onnx/csrc/offline-recognizer-funasr-nano-impl.cc

```diff
@@ -0,0 +1,492 @@
+  std::vector<int64_t> attention_mask(static_cast<size_t>(max_seq_len), 0);
+  // Copy text embeddings into inputs_embeds buffer
+  if (te_fp16) {
+    const uint16_t *p = text_embeds.GetTensorData<uint16_t>();
+    std::copy(p, p + static_cast<size_t>(context_len) * hidden_size,
+              inputs_embeds_fp16.data());
+  } else {
+    const float *p = text_embeds.GetTensorData<float>();
+    for (int64_t i = 0; i < static_cast<int64_t>(context_len) * hidden_size;
+         ++i) {
+      inputs_embeds_fp16[static_cast<size_t>(i)] = Fp32ToFp16(p[i]);
+    }
+  }
+  // Inject audio embeddings into inputs_embeds at the position of audio tokens
+  auto enc_info2 = encoder_out.GetTensorTypeAndShapeInfo();
+  auto enc_et =
+      static_cast<ONNXTensorElementDataType>(enc_info2.GetElementType());
+  int32_t copy_len = std::min(fake_token_len, audio_token_len);
+  if (enc_et == ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT16) {
+    const uint16_t *enc = encoder_out.GetTensorData<uint16_t>();
+    for (int32_t t = 0; t < copy_len; ++t) {
+      const uint16_t *src = enc + static_cast<int64_t>(t) * hidden_size;
+      uint16_t *dst = inputs_embeds_fp16.data() +
+                      static_cast<int64_t>(fbank_beg_idx + t) * hidden_size;
+      std::copy(src, src + hidden_size, dst);
+    }
+  } else if (enc_et == ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT) {
+    const float *enc = encoder_out.GetTensorData<float>();
+    for (int32_t t = 0; t < copy_len; ++t) {
+      const float *src = enc + static_cast<int64_t>(t) * hidden_size;
+      uint16_t *dst = inputs_embeds_fp16.data() +
+                      static_cast<int64_t>(fbank_beg_idx + t) * hidden_size;
+      for (int32_t d = 0; d < hidden_size; ++d) {
+        dst[d] = Fp32ToFp16(src[d]);
+      }
+    }
+  } else {
+    SHERPA_ONNX_LOGE("encoder_out elem_type=%d not supported", (int)enc_et);
+    result.text = "";
+    return result;
+  }
+  // Set attention mask for context tokens
+  for (int32_t i = 0; i < context_len; ++i) attention_mask[i] = 1;
+  int32_t valid_len = context_len;
+  std::vector<int64_t> generated_ids;
+  
+  generated_ids.reserve(funasr_config.max_new_tokens);
+  const int64_t eos_id = tokenizer_->GetEosTokenId();
+  const int64_t im_end_id = tokenizer_->GetImEndTokenId();
+  const int32_t max_new_tokens =
+      funasr_config.max_new_tokens > 0 ? funasr_config.max_new_tokens : 256;
+  std::vector<std::pair<Ort::Value, Ort::Value>> past_key_values;
+  bool is_first_step = true;
+  // Autoregressive generation loop
+  for (int32_t step = 0; step < max_new_tokens; ++step) {
+    if (valid_len >= max_seq_len) break;
+    Ort::Value logits(nullptr);
+    
+    if (is_first_step) {
+      // First step: use prefill model with full context
+      std::array<int64_t, 3> embeds_shape{1, context_len, hidden_size};
+      Ort::Value inputs_embeds_tensor = Ort::Value::CreateTensor(
+          memory_info, inputs_embeds_fp16.data(),
```

Comment:
Could you explain how you obtain the fp16 model?

Do we need to handle fp16 inputs and outputs?

You might want to take a look at the following script:
https://github.com/k2-fsa/sherpa-onnx/blob/acd44bf9ab0ee83c2189c93b442038c0b4e66786/scripts/uvr_mdx/add_meta_data_and_quantize.py#L31

If you set keep_io_types=True, the model’s input and output will always remain float32.